spikes/lk_sensors_imu_lsm6dsox_fusion_calibration/main.cpp

```diff
@@ -0,0 +1,143 @@
+// Leka - LekaOS
+// Copyright 2022 APF France handicap
+// SPDX-License-Identifier: Apache-2.0
+
+#include <cinttypes>
+
+#include "drivers/HighResClock.h"
+#include "rtos/ThisThread.h"
+
+#include "CoreI2C.h"
+#include "CoreLSM6DSOX.hpp"
+#include "LogKit.h"
+#include "fusion/Fusion.h"
+
+using namespace std::chrono;
+using namespace leka;
+
+namespace {
+
+namespace imu {
+
+	namespace internal {
+
+		auto drdy_irq = CoreInterruptIn {PinName::SENSOR_IMU_IRQ};
+		auto i2c	  = CoreI2C(PinName::SENSOR_IMU_TH_I2C_SDA, PinName::SENSOR_IMU_TH_I2C_SCL);
+
+	}	// namespace internal
+
+	CoreLSM6DSOX lsm6dsox(internal::i2c, internal::drdy_irq);
+
+}	// namespace imu
+
+namespace fusion {
+
+	constexpr auto kODR_HZ = int {52};
+
+	auto ahrs = FusionAhrs {};
+
+	const FusionAhrsSettings settings = {
+		.gain				   = 0.5F,
+		.accelerationRejection = 10.0F,
+		.magneticRejection	   = 0.0F,
+		.rejectionTimeout	   = static_cast<unsigned int>(5 * kODR_HZ),   // ? # of samples in 5 seconds
+	};
+
+	auto timestamp		   = rtos::Kernel::Clock::now();
+	auto previousTimestamp = rtos::Kernel::Clock::now();
+
+	auto global_offset = FusionOffset {};
+
+	constexpr auto CALIBRATION = bool {true};
+	// constexpr auto CALIBRATION = bool {false};
+
+	void callback(const interface::LSM6DSOX::SensorData &data)
+	{
+		timestamp  = rtos::Kernel::Clock::now();
+		auto ts_us = mbed::HighResClock::now().time_since_epoch().count();
+
+		// ? Acquire latest sensor data
+		auto gyroscope	   = FusionVector {{data.gy.x, data.gy.y, data.gy.z}};
+		auto accelerometer = FusionVector {{data.xl.x, data.xl.y, data.xl.z}};
+
+		if constexpr (CALIBRATION) {
+			// ? Define calibration offsets
+			// ? Data: https://www.dropbox.com/scl/fi/cue7qpb77892rozyjbmcq/2022_01_16-IMU-Calibration_Data.xlsx
+			// const auto gyroscopeOffset = FusionVector {{}};
+			// const auto accelerometerOffset = FusionVector {{}};
+			constexpr auto gyroscopeOffset	   = FusionVector {{0.02544522554, -0.3286247803, 0.3205770357}};
+			constexpr auto accelerometerOffset = FusionVector {{0.006480437024F, -0.01962820621F, 0.003259031049F}};
+
+			// ? Apply calibration offsets
+			gyroscope =
+				FusionCalibrationInertial(gyroscope, FUSION_IDENTITY_MATRIX, FUSION_VECTOR_ONES, gyroscopeOffset);
+			accelerometer = FusionCalibrationInertial(accelerometer, FUSION_IDENTITY_MATRIX, FUSION_VECTOR_ONES,
+													  accelerometerOffset);
+
+			// ? Update gyroscope offset correction algorithm
+			gyroscope = FusionOffsetUpdate(&global_offset, gyroscope);
```

Comment:
I read the documentation and examples from github Fusion, but I still don't understand the difference between `FusionCalibrationInertial` and `FusionOffsetUpdate`.

Reply:
- `FusionCalibrationInertial` - see https://github.com/xioTechnologies/Fusion#sensor-calibration - it's just a calibration function, you apply offset to the raw values to have the correct ones. The offset values are calculated before running the application. 

- `FusionOffsetUpdate` - see https://github.com/xioTechnologies/Fusion#gyroscope-offset-correction-algorithm - this is an inflight calibration algorithm: it senses when the device is stationary and fine tunes the offsets of the gyroscope